brainstorming.md

```diff
@@ -0,0 +1,508 @@
+- `user-indexed` - each row is addressable (indexed) by one or more user-specified index columns, which must be named columns in the dataframe schema.
+
+Row-indexed datafames must have no user-specified index columns and at least one column defined in their schema. User-indexed dataframes must have at least one schema-defined column named as an index column. The user-specified schema may not redefine the `__rowid` pseudo-column.
+
+Most language-specific bindings will provide convertors between SOMADataFrame and other convenient data structures, such as Python `pandas.DataFrame`, R `data.frame`.
+
+> ℹ️ **Note** - on a TileDB implementation, `row-indexed` SOMADataFrames will likely map to a _dense_ array, with a single uint64 dimension called **rowid, with domain [0 #nrows). A `user-indexed` SOMADataframe will be a _sparse_ array, with a dimension for each user-specified index column, and a hidden **rowid attribute.
+
+### SOMADenseNdArray
+
+`SOMADenseNdArray` is a monomorphic, dense, N-dimensional array with offset (zero-based) integer indexing on each dimension. The `SOMADenseNdArray` has a user-defined schema, which includes:
+
+- type - a `primitive` type, expressed as an Arrow type (eg, `int64`, `float32`, etc)
+- shape - the shape of the array, ie, number and length of each dimension
+
+All dimensions must have a positive, non-zero length.
+
+> ℹ️ **Note** - on TileDB this is an dense array with `N` uint64 dimensions of domain [0, maxUint64), and a single attribute.
+
+### SOMASparseNdArray
+
+`SOMASparseNdArray` is a monomorphic, sparse, N-dimensional array with offset (zero-based) integer indexing on each dimension. The `SOMASparseNdArray` has a user-defined schema, which includes:
+
+- type - a `primitive` type, expressed as an Arrow type (eg, `int64`, `float32`, etc)
+- shape - the shape of the array, ie, number and length of each dimension
+
+All dimensions must have a positive, non-zero length.
+
+> ℹ️ **Note** - on TileDB this is an sparse array with `N` uint64 dimensions of domain [0, maxUint64), and a single attribute.
+
+## Composed Types
+
+Composed types are defined as a composition of foundational types, adding name, type and indexing constraints. These types are intended to facilitate data interoperability, ease of use, and _potentially_ enable implementation optimizations by virtue of their typing and structural guarantees. The initial composed types are motivated by single cell biology, but additional types may be added in the future for more diverse use cases.
+
+### SOMAExperiment & SOMAMeasurements
+
+`SOMAExperiment` is a specialized `SOMACollection`, representing an annotated 2-D matrix of measurements. In the single-cell biology use case, a `SOMAExperiment` can represent multiple modes of measurement across a single collection of cells. Within a `SOMAExperiment`, a set of measurements on a single set of variables (features) is represented as a `SOMAMeasurements`.
```

Comment:
perhaps nitty: `SOMAMeasurements` already being plural is awkward. in the presence of multiple assays, there are ... SOMAMeasurementses? i'd love for us to have a word which is singular for one dict-slot of `SOMAExperiment`. Whether `SOMAMeasurement`, `SOMAAssay`, `SOMAMeasuremntSet`, `SOMAMeasurementItem`, or something else -- something singular.

Reply:
I defer to @ambrosejcarr who owns naming!

The original proposal was `SOMAMesurementSet`, which satisfies your criteria.  I am happy to adopt that.  The change to `SOMAMesurements` was solely to recognize that the type is not a true set in the mathematical or data structure sense of the word.

@johnkerl - do you think `SOMAMeasurement` or `SOMAMeasurementSet` is best (those are the two I'm comfortable with)?  I do not think `SOMAAssay` is good - it overloads the meaning with unnecessary specificity IMHO.

Reply:
sold - will make this change.

Reply:
I didn't have strong feelings between `SOMAMeasurement` and `SOMAMeasurementSet`. I find the argument against `Set` compelling, so am happy with the former.